larray/core/session.py

```diff
@@ -1392,6 +1393,155 @@ def display(k, v, is_metadata=False):
         return res
 
 
+# XXX: I wonder if we shouldn't create an AbstractSession instead of defining the _disabled()
+#      private method below.
+#      Auto-completion on any instance of a class that inherits from FrozenSession
+#      should not propose the add(), update(), filter(), transpose() and compact() methods
+class FrozenSession(Session):
+    """
+    The purpose of the present class is to be inherited by user defined classes where parameters
+    and variables of a model are defined (see examples below). These classes will allow users to
+    benefit from the so-called 'autocomplete' feature from development software such as PyCharm
+    (see the Notes section below) plus the main features of the :py:obj:`Session()` objects.
+
+    After creating an instance of a user defined 'session', some restrictions will be applied on it:
+
+    - **it is not possible to add or remove any parameter or variable,**
+    - **all non array variables (axes, groups, ...) cannot be modified,**
+    - **only values of array variables can be modified, not their axes.**
```

Comment:
Honestly, I am unsure about the "more axes" part. Personally, I would certainly like to work like that (declare all required axes by my model and let it work with any dataset which has at least those axes), but then you lose the quick reference we have now were we can quickly lookup the array name and know its exact axes, instead of knowing minimum required axes you can rely on. Knowing exact axes without running the model has some value too. If I was a modeller, generic code would be more important than "precise" axes, but I wonder if our users wouldn't favour precise axes over generic/powerful code.